appinventor/components-ios/src/Trendline.swift

```diff
@@ -0,0 +1,592 @@
+  private var exponentialRegression: ExponentialRegression = ExponentialRegression()
+  private var logarithmicRegression: LogarithmicRegression = LogarithmicRegression()
+  private var currentModel: TrendlineCalculator?
+  private var lastResults: [String: Any] = [:]
+  private var initialized: Bool = false
+  private var dataModel: ChartDataModel?
+  private var minX: Double = Double.infinity
+  private var maxX: Double = -Double.infinity
+  private var _density: CGFloat
+
+  @objc public init(_ chartContainer: Chart) {
+    self._container = chartContainer
+    self._density = UIScreen.main.scale
+    super.init()
+    self._container.addDataComponent(self)
+    self._dashed = createDashedPattern(density: _density)
+    self._dotted = createDottedPattern(density: _density)
+    self.currentModel = regression
+  }
+
+  @objc func Initialize() {
+    initialized = true
+    if dataModel == nil {
+      initChartData()
+    }
+  }
+
+  private func createDashedPattern(density: CGFloat) -> [CGFloat] {
+    let dashLength = 10 * density
+    let dashGap = 10 * density
+    return [dashLength, dashGap]
+  }
+
+  private func createDottedPattern(density: CGFloat) -> [CGFloat] {
+    let dotLength = 2 * density
+    let dotGap = 10 * density
+    return [dotLength, dotGap]
+  }
+
+  func draw(context: CGContext) {
+    switch _strokeStyle {
+    case .Dashed:
+      context.setLineDash(phase: 0, lengths: _dashed ?? [])
+    case .Dotted:
+      context.setLineDash(phase: 0, lengths: _dotted ?? [])
+    case .Solid:
+      context.setLineDash(phase: 0, lengths: [])
+    default:
+      break
+    }
+  }
+
+  func onDataSourceValueChange(_ component: DataSource, _ key: String?, _ newValue: AnyObject?) {
+    lastResults.removeAll()
+
+    guard let chartDataEntries = _chartData?.chartDataModel?._entries as? [DGCharts.ChartDataEntry] else {
+      print("No entries in the data source")
+      return
+    }
+
+    var x: [Double] = []
+    var y: [Double] = []
+
+    for entry in chartDataEntries {
+      let xValue = entry.x
+      let yValue = entry.y
+
+      x.append(xValue)
+      y.append(yValue)
+    }
+
+
+    guard !x.isEmpty, x.count >= 2, x.count == y.count else {
+      print("Not enough entries in the data source or unequal X and Y data points")
+      return
+    }
+
+    if let currentModel = currentModel {
+      let computedResults = currentModel.compute(x: x, y: y)
+
+      // Now you can use computedResults directly
+      // For example, updating the state or refreshing the UI based on these results
+      lastResults = computedResults
+
+      if initialized {
+        DispatchQueue.main.async {
+          self._container.chartView?.refresh()
+        }
+      }
+    } else {
+      print("Error: Current model is not set")
+    }
+  }
+
+
+  func onReceiveValue(_ component: RealTimeDataSource, _ key: String?, _ value: AnyObject?) {
+
+  }
+
+  @objc public var ChartData: ChartData2D {
+    get {
+      return _chartData!
```

Comment:
Be careful with the forced unwrapping here since if for some reason _chartData is nil when this is called it will crash the app.